File: go/3121. Count the Number of Special Characters II.go
func numberOfSpecialChars(word string) int {
	lower := [26]int{}
	upper := [26]int{}
	lastLower := [26]int{}
	firstUpper := [26]int{}
	for i := range 26 {
		lastLower[i] = -1
		firstUpper[i] = -1
	}
	for i, c := range word {
		if c >= 'a' && c <= 'z' {
			lower[c-'a']++
			lastLower[c-'a'] = i
		} else if c >= 'A' && c <= 'Z' {
			upper[c-'A']++
			if firstUpper[c-'A'] == -1 {
				firstUpper[c-'A'] = i
			}
		}
	}
	answer := 0
	for i := range 26 {
		if lower[i] > 0 && upper[i] > 0 && lastLower[i] < firstUpper[i] {
			answer++
		}
	}
	return answer
}

File: python/3121. Count the Number of Special Characters II.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        lower = [0] * 26
        upper = [0] * 26
        last_lower = [-1] * 26
        first_upper = [-1] * 26
        for i in range(len(word)):
            if word[i].islower():
                lower[ord(word[i]) - ord('a')] += 1
                last_lower[ord(word[i]) - ord('a')] = i
            else:
                upper[ord(word[i]) - ord('A')] += 1
                if first_upper[ord(word[i]) - ord('A')] == -1:
                    first_upper[ord(word[i]) - ord('A')] = i
        return sum([1 for i in range(26) if lower[i] > 0 and upper[i] > 0 and last_lower[i] < first_upper[i]])
        